File: ruby/exercises/space-age/space_age.rb
# SpaceAge
class SpaceAge
  ORBITAL_PERIODS = {
    'mercury' => 7_600_530.24,
    'venus' =>   19_413_907.2,
    'earth' =>   31_558_149.76,
    'mars' =>    59_354_294.4,
    'jupiter' => 374_335_776.0,
    'saturn' =>  929_596_608.0,
    'uranus' =>  2_661_041_808.0,
    'neptune' => 5_200_418_592.0
  }.freeze

  attr_reader :seconds
  def initialize(seconds)
    @seconds = seconds
  end

  def on_planet(planet)
    (seconds / ORBITAL_PERIODS[planet]).round(2)
  end

  ORBITAL_PERIODS.keys.each do |planet|
    define_method :"on_#{planet}" do
      on_planet(planet)
    end
  end
end

File: go/space-age/space_age.go
package space

// Planet names
type Planet string

const secondsInEarthYear = 31557600

var orbitalPeriods = map[Planet]float64{
	"Mercury": 0.2408467,
	"Venus":   0.61519726,
	"Mars":    1.8808158,
	"Jupiter": 11.862615,
	"Saturn":  29.447498,
	"Uranus":  84.016846,
	"Neptune": 164.79132,
}

// Age returns the age in years of someone on planet given their age in seconds.
func Age(ageInSeconds float64, planet Planet) float64 {
	if planet == "Earth" {
		return ageOnEarth(ageInSeconds)
	}
	return ageOnEarth(ageInSeconds) / orbitalPeriods[planet]
}

// AgeOnEarth returns the age of somone on Earth given their age in seconds.
func ageOnEarth(ageInSeconds float64) float64 {
	return ageInSeconds / secondsInEarthYear
}
